fl_training.py

```diff
@@ -149,191 +194,169 @@ def compute_test_score(
-                        [
-                            np.array(weights_).mean(axis=0)
-                            for weights_ in zip(*weights_list_tuple)
-                        ]
-                    )
-
-                aggregated_model = utils.generate_new_cnn_model()
-                aggregated_model.set_weights(new_weights)
-                aggregated_model.compile(
-                    loss=keras.losses.categorical_crossentropy,
-                    optimizer="adam",
-                    metrics=["accuracy"],
-                )
+                node_model = build_aggregated_model(aggregate_model_weights(model_list, aggregation_weights))
 
-                # Evaluate model for early stopping, on a central and dedicated 'early stopping validation' set
-                model_evaluation = aggregated_model.evaluate(
-                    x_val_global,
-                    y_val_global,
-                    batch_size=constants.BATCH_SIZE,
-                    verbose=0,
-                )
-                current_val_loss = model_evaluation[0]
-                global_val_acc.append(model_evaluation[1])
-                global_val_loss.append(current_val_loss)
-
-                # Early stopping
-                print("   Checking if early stopping critera are met:")
-                if is_early_stopping:
-                    # Early stopping parameters
-                    if (
-                        epoch >= constants.PATIENCE
-                        and current_val_loss > global_val_loss[-constants.PATIENCE]
-                    ):
-                        print("      -> Early stopping critera are met, stopping here.")
-                        break
-                    else:
-                        print(
-                            "      -> Early stopping critera are not met, continuing with training."
-                        )
-
-            # Training phase
-            val_acc_list = []
-            acc_list = []
+            # Iterate over nodes for training each individual model
             for node_index, node in enumerate(node_list):
 
-                print("   Training on node " + str(node))
-                node_model = utils.generate_new_cnn_model()
-
-                # Model weights are the averaged weights
-                if not is_first_epoch:
-                    node_model.set_weights(new_weights)
-                    node_model.compile(
-                        loss=keras.losses.categorical_crossentropy,
-                        optimizer="adam",
-                        metrics=["accuracy"],
-                    )
+                print("   Training on node " + str(node_index) + " - " + str(node))
 
-                # Train on whole node local data set
+                # Train on node local data set
                 history = node_model.fit(
-                    node.x_train,
-                    node.y_train,
+                    minibatched_x_train[node_index][minibatch+1],
```

Comment:
Why `+1`?

Reply:
Because the first piece of the split is empty, by construction.

Reply:
Ok, clarified